metricbeat/scripts/mage/docsCollector.go

```diff
@@ -0,0 +1,488 @@
+
+For a description of each field in the metricset, see the
+<<exported-fields-%s,exported fields>> section.
+
+`
+
+// get the release tag. This kind of logic is needed because any
+// module/metricset missing a release is categorized as "experimental"
+// There are three valid tags: experimental, beta, ga
+func getRelease(rel string) (string, error) {
+	if rel != "" {
+		if rel != "beta" && rel != "ga" {
+			return "", fmt.Errorf("unknown release tag %s", rel)
+		}
+		return rel, nil
+	}
+	return "experimental", nil
+}
+
+//create the path
+func createDocsPath(module string) error {
+	return os.MkdirAll(mage.OSSBeatDir(filepath.Join("docs", module)), 0755)
+}
+
+//test for a `_meta/docs.asciidoc` in a given directory
+func testIfDocsInDir(moduleDir string) (bool, error) {
+	moduledir, err := os.Stat(moduleDir)
+	if err != nil {
+		return false, err
+	}
+	if moduledir.Mode().IsRegular() {
+		return false, nil
+	}
+	_, err = os.Stat(filepath.Join(moduleDir, "_meta/docs.asciidoc"))
+	if os.IsNotExist(err) {
+		return false, nil
+	} else if err != nil {
+		return false, errors.Wrap(err, "error looking for asciidoc")
+	}
+	return true, nil
+}
+
+//load the module-specific fields.yml file
+func loadModuleFields(file string) (moduleData, error) {
+	fd, err := ioutil.ReadFile(file)
+	if err != nil {
+		return moduleData{}, errors.Wrap(err, "failed to read from spec file")
+	}
+	//Cheat and use the same struct.
+	var mod []moduleData
+	if err = yaml.Unmarshal(fd, &mod); err != nil {
+		return mod[0], err
+	}
+	module := mod[0]
+
+	rel, err := getRelease(module.Release)
+	if err != nil {
+		return mod[0], err
+	}
+	module.Release = rel
+
+	return module, nil
+}
+
+//The doc generator only needs the release field from the metricset-level fields.yml.
+func getReleaseState(metricsetPath string) (string, error) {
+	raw, err := ioutil.ReadFile(metricsetPath)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read from spec file")
+	}
+
+	type metricset struct {
+		Release string `yaml:"release"`
+	}
+	var rel []metricset
+	if err = yaml.Unmarshal(raw, &rel); err != nil {
+		return "", err
+	}
+
+	return getRelease(rel[0].Release)
+}
+
+//check to see if the metricset has dashboards
+func hasDashboards(modulePath string) bool {
+	_, err := os.Stat(filepath.Join(modulePath, "_meta/kibana"))
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+//use the reference file if it exists. if not, the normal one.
+func getConfigfile(modulePath string) (string, error) {
+	var cfgFile string
+	cfg := filepath.Join(modulePath, "_meta/config.reference.yml")
+	_, err := os.Stat(cfg)
+	if err == nil {
+		cfgFile = cfg
+	} else if os.IsNotExist(err) {
+		cfgBack := filepath.Join(modulePath, "_meta/config.yml")
+		if _, err := os.Stat(cfgBack); err != nil {
+			return "", fmt.Errorf("config file in %s could not be found", modulePath)
+		}
+		cfgFile = cfgBack
+	} else {
+		return "", fmt.Errorf("config file in %s could not be found", modulePath)
+	}
```

Comment:
Consider implementing this as a loop trying with both options, I think it can be more readable and it will lbe more more easy to modify if we change these config files. In ~pseudocode:
```
paths := []string{"_meta/config.reference.yml", "_meta/config.yml"}
for path in paths {
    path = filepath.Join(modulePath, path)
    if not path exists {
       continue
    }
    return ioutil.ReadFile(path)
}
return "", fmt.Errorf("config file in %s could not be found", modulePath)
```

Reply:
Ah, that makes way more sense. I was kinda banging my head against a wall trying to find a less ugly way to do that.